Program 'operator' class methods.

rxu opened the pull request:
`TODO` comments have been retained intentionally so far.

rxu committed:
Program 'operator' class methods.

rxu commented:
!set WIP

iMattPro commented on operator/operator.php:
```diff
@@ -70,6 +82,19 @@ public function set_user_categories($forum_id)
+		if ($this->user->data['is_registered'])
+		{
+			$sql = 'UPDATE ' . USERS_TABLE . '
+				SET collapsible_categories = ' . $this->db->sql_escape(serialize($this->collapsed_categories)) . '
```
Quotes should be added in the stmt around the sql_escaped string

iMattPro commented on operator/operator.php:
```diff
@@ -70,6 +82,19 @@ public function set_user_categories($forum_id)
+			$sql = 'UPDATE ' . USERS_TABLE . '
+				SET collapsible_categories = ' . $this->db->sql_escape(serialize($this->collapsed_categories)) . '
+				WHERE user_id = ' . (int) $this->user->data['user_id'];
+			$this->db->sql_query($sql);
```
This function should return false when no data was set (helps the AJAX response complete only when changes were successful) so there should be:

``` php
// There was an error updating the user's data
if (!$this->db->sql_affectedrows())
{
    return false;
}
```

rxu replied:
What would cause that?

iMattPro replied:
Any possible error that might occur, no matter how rare. We only want to return true if the update actually occurs.

rxu replied:
Do we return false regardless of the `set_cookie_categories()` result or should we skip the cookie part in this case? I'm confused a bit here.

iMattPro replied:
Yes we will because this would only happen in the event of some kind of error with the user. Otherwise their cookies and db would get "out of sync" with each other.

We always want to set the reg users db and cookie together, or not at all.

iMattPro commented on operator/operator.php:
```diff
@@ -70,6 +82,19 @@ public function set_user_categories($forum_id)
+			$this->db->sql_query($sql);
+		}
+
+		if (!$this->user->data['is_bot'])
```
I would put this is_bot check at the top of the method and just return false.

Then this set_cookie_cats can be left outside of a conditional and it should return a value t/f, ie:
`return $this->set_cookie_categories($forum_id);`

iMattPro commented on operator/operator.php:
```diff
@@ -92,6 +127,8 @@ public function set_cookie_categories($forum_id)
 		// TODO
 		// 1.update the cookie with json_encoded array of collapsed category data ($this->collapsed_categories)
+
+		$this->user->set_cookie('ccat', json_encode($this->collapsed_categories), strtotime('+1 year'));
```
Lets add a `return true` at the end (this method needs to return a boolean)

iMattPro commented on operator/operator.php:
```diff
@@ -80,6 +105,16 @@ public function get_cookie_categories()
+
+		$cookie_data = '';
+		$cookie_categories = array();
+		if ($this->request->is_set($this->config['cookie_name'] . '_ccat', \phpbb\request\request_interface::COOKIE))
```
I don't think we need to check if the cookie is set. The request_variable below will return `''` if none is found. I also don't think the variables need initializing above. We should also add some comments. So just:

``` php
$cookie_data = htmlspecialchars_decode($this->request->variable($this->config['cookie_name'] . '_ccat', '', true, \phpbb\request\request_interface::COOKIE));

return (array) json_decode($cookie_data, true);
```

iMattPro commented on operator/operator.php:
```diff
@@ -57,6 +57,18 @@ public function get_user_categories()
 		// 2.if no categories found, call get_cookie_categories()
 		// 3.return the categories or an empty array
+
+		$collapsible_categories = array();
```
This is not needed, as it is set in one of the conditions below

iMattPro commented on operator/operator.php:
```diff
@@ -80,6 +105,16 @@ public function get_cookie_categories()
+		if ($this->request->is_set($this->config['cookie_name'] . '_ccat', \phpbb\request\request_interface::COOKIE))
+		{
+			$cookie_data = htmlspecialchars($this->request->variable($this->config['cookie_name'] . '_ccat', '', true, \phpbb\request\request_interface::COOKIE));
+			$cookie_categories = (array) json_decode($cookie_data);
```
json_decode needs a true for second parameter so it returns and array instead of an object

iMattPro commented on operator/operator.php:
```diff
@@ -80,6 +105,16 @@ public function get_cookie_categories()
+		$cookie_categories = array();
+		if ($this->request->is_set($this->config['cookie_name'] . '_ccat', \phpbb\request\request_interface::COOKIE))
+		{
+			$cookie_data = htmlspecialchars($this->request->variable($this->config['cookie_name'] . '_ccat', '', true, \phpbb\request\request_interface::COOKIE));
```
Should be htmlspecialchars_decode

iMattPro commented on operator/operator.php:
```diff
@@ -57,6 +57,18 @@ public function get_user_categories()
+		$collapsible_categories = array();
+		if (!empty($this->user->data['collapsible_categories']))
+		{
+			$collapsible_categories = (array) unserialize($this->user->data['collapsible_categories']);
```
This won't really result in an empty array which is what we want to return if unserialize has a problem. This would return `array(0 => false)`

rxu committed:
Code fixes and minor optimizations.
- (un)serialize -> json encode/decode.
- changes to conform interface class.
- add comments.

iMattPro commented on operator/operator.php:
```diff
@@ -53,10 +53,20 @@ public function __construct(\phpbb\config\config $config, \phpbb\db\driver\drive
-		// 3.return the categories or an empty array
+		// Get the categories by json_decode() the user object data 'collapsible_categories'
+		// If no categories found in the database, call get_cookie_categories()
+		// Return the categories array or an empty array if there were not categories found
```
We don't need so much commenting, since the interface docblock also explains some of this. We can simplify like:

``` php
// Get categories from the user object
if (!empty($this->user->data['collapsible_categories']))
{
    // Data needs to be json decoded and cast as an array
    $collapsible_categories = (array) json_decode($this->user->data['collapsible_categories'], true);
}
else
{
    // The user object had no categories, check for a cookie
    $collapsible_categories = $this->get_cookie_categories();
}
```

iMattPro commented on operator/operator.php:
```diff
@@ -64,12 +74,36 @@ public function get_user_categories()
-		// Set the collapsed category data array
-		$this->set_collapsed_categories($forum_id);
+		// Do not apply the ext functionality for bots
+		if (!$this->user->data['is_bot'])
```
The bot protection is not needed here. In fact what we should do is when we switch over to template events (pending the event being added to the core) we should check if is a bot in the template event, and not display the buttons/links at all there instead :smile: I will add that note to the issue for that.

Also, just FYI, if we were to do something like this, rather than wrapping the whole method in a condition, I prefer this style:

```
if (something went wrong)
{
    return false;
}

// continue with normal function design and return values...
```

iMattPro commented on operator/operator.php:
```diff
@@ -64,12 +74,36 @@ public function get_user_categories()
+			}
+
+			// If there was an error updating the user's data, return false
+			// Otherwise update cookie data as well
```
We can just say: `// There was an error updating the users data`

iMattPro commented on operator/operator.php:
```diff
@@ -64,12 +74,36 @@ public function get_user_categories()
+			}
+			else
+			{
+				$this->set_cookie_categories($forum_id);
```
This should not be in an else. This should always run. We want users to have both their db and cookie updated.

It should be where your `return true` is, but like::

```
// Set a cookie with the collapsed forum data
return $this->set_cookie_categories($forum_id);
```

iMattPro commented on operator/operator.php:
```diff
@@ -64,12 +74,36 @@ public function get_user_categories()
+
+			// If there was an error updating the user's data, return false
+			// Otherwise update cookie data as well
+			if (!$this->db->sql_affectedrows())
```
This should be inside the`if ($this->user->data['is_registered'])` condition, after the SQL query

```
if (!$this->db->sql_affectedrows())
{
    return false;
}
```

iMattPro commented on operator/operator.php:
```diff
@@ -64,12 +74,36 @@ public function get_user_categories()
+				$this->set_cookie_categories($forum_id);
+			}
+
+			return true;
```
In case those notes of mine were messy, basically, this is what I was looking for in this method:

``` php
// Set the collapsed category data array
$this->set_collapsed_categories($forum_id);

// Update the user if they are a member
if ($this->user->data['is_registered'])
{
    $sql = 'UPDATE ' . USERS_TABLE . "
        SET collapsible_categories = '" . $this->db->sql_escape(json_encode($this->collapsed_categories)) . "'
        WHERE user_id = " . (int) $this->user->data['user_id'];
    $this->db->sql_query($sql);

    // There was an error updating the user's data
    if (!$this->db->sql_affectedrows())
    {
        return false;
    }
}

// Set a cookie with the collapsed forum data
return $this->set_cookie_categories($forum_id);
```

iMattPro commented on operator/operator.php:
```diff
@@ -77,9 +111,13 @@ public function set_user_categories($forum_id)
-		// 2.return categories or an empty array
+		// Get categories from the cookie (htmlspecialchars_decode and json_decode apply)
+		// Return the categories array or an empty array if there were not categories found
+		$cookie_data = htmlspecialchars_decode($this->request->variable($this->config['cookie_name'] . '_ccat', '', true, \phpbb\request\request_interface::COOKIE));
```
Just a 1-liner comment please:
// Get categories from the cookie and htmlspecialchars decode it

iMattPro commented on operator/operator.php:
```diff
@@ -77,9 +111,13 @@ public function set_user_categories($forum_id)
+
+		// Cast the result to array to get an empty array if no categories were found
+		// or convert the cookie data to array if something's went wrong with it
+		return (array) json_decode($cookie_data, true);
```
1 line comment here too:
// json decode the cookie data and return an array

iMattPro commented on operator/operator.php:
```diff
@@ -90,8 +128,11 @@ public function set_cookie_categories($forum_id)
 
-		// TODO
-		// 1.update the cookie with json_encoded array of collapsed category data ($this->collapsed_categories)
+		// Update the cookie with json_encoded array of collapsed category data
```
`json_encoded` should not have underscore in this comment (my bad?)

rxu committed:
Further cleanup the code and comments.

rxu commented:
Updated #15.

iMattPro commented:
@rxu Looks good. Let me know if this is no longer WIP
